### PR TITLE
Dataset speed up read

### LINKED ISSUE #17353: Slow file listing performance
The download time of the inference nightly test dominates, and is very slow: {'download_time': 111.62019515037537, 'preprocess_time': 11.225638628005981, 'inference_time': 33.34679889678955, 'total_time': 156.1926326751709}.

https://buildkite.com/ray-project/periodic-ci/builds/654#4cb0b2ec-112f-48f3-b8a7-df8928443415

We should time how long it takes to list the files using boto3 directly, which would set a lower bound. I believe the current `read_binary_files` is very slow since it's doing extra IO operations (e.g., opening every file serially on the driver during file resolution).

cc @clarkzinzow

### PR DESCRIPTION
**Pull request by wuisawesome**
Closes https://github.com/ray-project/ray/issues/17353

### CODE REVIEW

File: python/ray/experimental/data/read_api.py [hunk 1]
```diff
@@ -103,21 +103,28 @@ def range_arrow(n: int, parallelism: int = 200) -> Dataset[ArrowRow]:
 @PublicAPI(stability="beta")
 def read_datasource(datasource: Datasource[T],
                     parallelism: int = 200,
+                    num_cpus: float = None,
```
Comment (ericl) [inline python/ray/experimental/data/read_api.py:106]: Please change this to just `remote_args` as in the Dataset .map() definition.

File: release/nightly_tests/dataset/inference.py [hunk 2]
```diff
@@ -75,17 +75,13 @@ def infer(batch):
...
-        "GPU", 400)))
-    time.sleep(5)
-
 start_time = time.time()
 
 print("Downloading...")
 ds = ray.experimental.data.read_binary_files(
-    "s3://anyscale-data/small-images/", parallelism=400)
-ds = ds.limit(100 * 1000)
+    "s3://anyscale-data/small-images/", parallelism=1000, num_cpus=0.05)
```
Comment (ericl) [inline release/nightly_tests/dataset/inference.py:82]: How about 0.2? 0.05 seems likely to cause instability.

**Review by ericl — CHANGES_REQUESTED** [review]

File: python/ray/experimental/data/read_api.py (hunk 1 above)
Comment (wuisawesome) [reply python/ray/experimental/data/read_api.py:106]: Do you mean to combine it with the existing `**read_args`? Or adding an argument `remote_args: Dict[str, Any]`?

Comment (ericl) [reply python/ray/experimental/data/read_api.py:106]: The latter, exactly as specified in .map().

Comment (wuisawesome) [reply python/ray/experimental/data/read_api.py:106]: Am I misunderstanding something? in .map, `**remote_args` is a kwargs field. In this function we already have `**read_args` as the kwargs, which would mean they are combined.

Comment (ericl) [reply python/ray/experimental/data/read_api.py:106]: I see, the issue is you can't have two kwargs fields. How about we make this a normal dict then (not kwargs).

**Comment by ericl** [comment]
=========================== short test summary info ============================
  | FAILED ::test_parquet_read - TypeError: unsupported operand type(s) for ** or...
  | FAILED ::test_convert_to_pyarrow - TypeError: unsupported operand type(s) for...
  | FAILED ::test_read_binary_files - TypeError: unsupported operand type(s) for ...
  | FAILED ::test_read_binary_files_with_fs - TypeError: unsupported operand type...
  | FAILED ::test_read_binary_files_with_paths - TypeError: unsupported operand t...
  | FAILED ::test_map_batch - TypeError: unsupported operand type(s) for ** or po...
  | FAILED ::test_json_read - TypeError: unsupported operand type(s) for ** or po...
  | FAILED ::test_zipped_json_read - TypeError: unsupported operand type(s) for *...
  | FAILED ::test_csv_read - TypeError: unsupported operand type(s) for *: 'int' ...
 

<br class="Apple-interchange-newline">

File: python/ray/experimental/data/read_api.py [hunk 3]
```diff
@@ -103,24 +103,30 @@ def range_arrow(n: int, parallelism: int = 200) -> Dataset[ArrowRow]:
 @PublicAPI(stability="beta")
 def read_datasource(datasource: Datasource[T],
                     parallelism: int = 200,
+                    remote_args: Dict[str, Any] = None,
```
Comment (ericl) [inline python/ray/experimental/data/read_api.py:106]:
```suggestion
                    ray_remote_args: Dict[str, Any] = None,
```

File: python/ray/experimental/data/read_api.py [hunk 4]
```diff
@@ -103,24 +103,30 @@ def range_arrow(n: int, parallelism: int = 200) -> Dataset[ArrowRow]:
...
                     **read_args) -> Dataset[T]:
     """Read a dataset from a custom data source.
 
     Args:
         datasource: The datasource to read data from.
         parallelism: The requested parallelism of the read.
         read_args: Additional kwargs to pass to the datasource impl.
+        remote_args: kwargs passed to @ray.remote in the read tasks.
```
Comment (ericl) [inline python/ray/experimental/data/read_api.py:114]:
```suggestion
        ray_remote_args: kwargs passed to ray.remote in the read tasks
            (e.g., {"num_cpus": 0.5} to increase read parallelism).
```

**Review by ericl — CHANGES_REQUESTED** [review]
Last comment

**Comment by wuisawesome** [comment]
Passing the dataset unit tests. I'll manually verify the nightly test performance after the wheels finish building.

**Review by ericl — APPROVED** [review]

**Pull request merged**